File: 110. 平衡二叉树/Solution.cpp
#include<bits/stdc++.h>

using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int traversal(TreeNode* node){
        if(node == NULL) return 0;
        int left = traversal(node->left);
        int right = traversal(node->right);
        if(left == -1 || right == -1 || abs(left-right) > 1){
            return -1;
        }
        return max(left, right) + 1;
    }
    bool isBalanced(TreeNode* root) {
        return traversal(root) == -1 ? false : true;
    }
};

File: 110. 平衡二叉树/Solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traversal(self, root):
        # 如果发现有子树不是平衡二叉树则逐级返回-1
        if not root:
            return 0
        left = self.traversal(root.left)
        right = self.traversal(root.right)
        if left == -1 or right == -1 or abs(left - right) > 1:
            return -1
        else:
            return max(left, right) + 1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self.traversal(root) != -1